### PR TITLE
Makes SplitMetaDataAndFooterOffsets versionable.

### LINKED ISSUE #786: Add a version in the split metadata
Right now we rely on the JSON serialization of the split metadata to store data in metastores.
This makes it possible to handle most of our migration on this JSON object and not in the N different metastore impelmentation.

On the other hand, it requires us to be able to read all older versions.
Right now we do not keep a version within this json.

### LINKED ISSUE #787: Have unit tests for track split metadata versions deserialization.

### PR DESCRIPTION
**Pull request by fulmicoton**
- adds a version to the SplitMetaDataAndFooterOffsets
    - adds backward compatibility testing with autogenerated data.
    - serializes tags in sorted order
    - isolated MetadataSet into the single file metastore
    - isolated split metadata related things to its own file.
    - isolated MetadataSet into the single file metastore

Closes #786
Closes #787

### CODE REVIEW

File: quickwit-backward-compat/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Backward compatibility test project.
+
+This is just a project used to test backward compatibility of Quickwit.
+Right now, only SplitMetadata is test.
+
+The build.rs script autocreates json sample files something changes in the format.
```
Comment (fmassot) [inline quickwit-backward-compat/README.md:6]:
```suggestion
The build.rs script autocreates json sample files when something changes in the format.
```

File: quickwit-backward-compat/build.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            tags: ["234".to_string(), "aaa".to_string()].into_iter().collect(),
+            demux_num_ops: 1,
+        },
+        footer_offsets: 1000..2000,
+    }
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let split_metadata = sample_split_metadata_for_regression();
+    let split_metadata_value = serde_json::to_value(&split_metadata)?;
```
Comment (PSeitz) [inline quickwit-backward-compat/build.rs:43]: reordering fields could cause creation of a new entry although the format is the same

File: quickwit-backward-compat/tests/split_metadata_non_regression.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+fn check_version(
+    version: usize,
+    split_metadata: &SplitMetadataAndFooterOffsets,
+) -> anyhow::Result<()> {
+    // Feel free to enrich this function if we start introducing default values that are not
+    // deterministic. (e.g. depends on time.)
+    static EXPECTED_DESERIALIZED_SPLIT_METADATA: OnceCell<
+        HashMap<usize, SplitMetadataAndFooterOffsets>,
+    > = OnceCell::new();
+    let expected_split_metadata = EXPECTED_DESERIALIZED_SPLIT_METADATA
```
Comment (PSeitz) [inline quickwit-backward-compat/tests/split_metadata_non_regression.rs:38]: I would remove the once cell and pass it as parameter from `test_backward_compatibility`

File: quickwit-indexing/src/actors/packager.rs [hunk 4]
```diff
@@ -17,7 +17,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-use std::collections::HashSet;
+use std::collections::BTreeSet;
```
Comment (fmassot) [inline quickwit-indexing/src/actors/packager.rs:20]: this change is driven by 2 things I guess:
- it's nice to sort tags
- we do iterations of this field

File: quickwit-metastore/src/metastore/single_file_metastore.rs [hunk 5]
```diff
@@ -28,19 +28,28 @@ use quickwit_storage::{
...
+    IndexMetadata, Metastore, MetastoreError, MetastoreFactory, MetastoreResolverError,
+    MetastoreResult, SplitMetadata, SplitMetadataAndFooterOffsets, SplitState,
 };
 
 /// Metadata file managed by [`SingleFileMetastore`].
 const META_FILENAME: &str = "quickwit.json";
 
+/// A MetadataSet carries an index metadata and its split metadata.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MetadataSet {
```
Comment (fmassot) [inline quickwit-metastore/src/metastore/single_file_metastore.rs:46]: nice idea to move that into the single file metastore

File: quickwit-metastore/src/split_metadata.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::ops::{Range, RangeInclusive};
+use std::str::FromStr;
+
+use chrono::Utc;
+use serde::{Deserialize, Serialize};
+
+/// Carries split and bundle offsets for single read metadata.
```
Comment (fmassot) [inline quickwit-metastore/src/split_metadata.rs:28]: by `for single read metadata`, do you mean that we should not mutate it?

File: quickwit-metastore/src/split_metadata.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            footer_offsets: split_metadata_and_footer_offsets.footer_offsets,
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for SplitMetadataAndFooterOffsets {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where D: serde::Deserializer<'de> {
+        let split_metadata_value = serde_json::Value::deserialize(deserializer)?;
+        // Unfortunately, it is not possiible to tell serde that in the absence
```
Comment (fmassot) [inline quickwit-metastore/src/split_metadata.rs:85]:
```suggestion
        // Unfortunately, it is not possible to tell serde that in the absence
```

File: quickwit-metastore/src/split_metadata.rs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        // of a tag, a given tag should be considered as the default.
+        // We did not populate in the past, and are therefore required to
```
Comment (fmassot) [inline quickwit-metastore/src/split_metadata.rs:87]: rephrasing proposal 
```suggestion
        // Old serialized metadata do not contain the version tag and therefore we are required to
```

**Review by fmassot — APPROVED** [review]
LGTM, code is very clear.
I added (very) minor comments.

File: quickwit-backward-compat/tests/split_metadata_non_regression.rs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+use std::fs;
+use std::path::Path;
+use std::str::FromStr;
+
+use anyhow::Context;
+use once_cell::sync::OnceCell;
+use quickwit_metastore::{SplitMetadata, SplitMetadataAndFooterOffsets, SplitState};
+
+fn check_version(
+    version: usize,
```
Comment (PSeitz) [inline quickwit-backward-compat/tests/split_metadata_non_regression.rs:30]: I don't get this, why do we store and compare per version number, the idea is to be able to have old versions be compatible with new versions?

Also relevant may be to be able read newer version, depending on how we roll out updates
(may be hard to test ;)

Comment (fulmicoton) [reply quickwit-backward-compat/tests/split_metadata_non_regression.rs:30]: Right now it works but that is dumb actually.

I think we will have one expected splitmetadata to fill per version. 
We could save it in another file or something like that.

File: quickwit-backward-compat/build.rs (hunk 2 above)
Comment (fulmicoton) [reply quickwit-backward-compat/build.rs:43]: That's fine. It's up to the dev to remove the file then.

File: quickwit-metastore/src/split_metadata.rs (hunk 6 above)
Comment (fulmicoton) [reply quickwit-metastore/src/split_metadata.rs:28]:
```suggestion
/// Carries split metadata and footer offsets. The footer offsets
/// make it possible to download the footer in a single call to `.get_slice(...)`.
```

File: quickwit-indexing/src/actors/merge_planner.rs [hunk 10]
```diff
@@ -138,8 +138,8 @@ mod tests {
         Some(time_start..=time_end)
     }
 
-    fn compute_merge_tags(splits: &[SplitMetadata]) -> HashSet<String> {
-        let mut tag_set: HashSet<String> = HashSet::new();
+    fn compute_merge_tags(splits: &[SplitMetadata]) -> BTreeSet<String> {
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-indexing/src/actors/merge_planner.rs:141]: To get the ordering

File: quickwit-backward-compat/tests/split_metadata_non_regression.rs (hunk 3 above)
Comment (fulmicoton) [reply quickwit-backward-compat/tests/split_metadata_non_regression.rs]: change the logic to check against a expected.json file.

**Pull request merged**